course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,447 @@
+
+ - not equal to `!=`
+ - less than `<`
+ - greater than `> `
+ - less than or equal to `<=`
+ - greater than or equal to `>=`
+
+Note that we do not use the `=` operator when it comes to specifying values we want to find in a column. This is an important point in python. The `==` operator compares the value or equality of two objects and is the right operator to use. It is a `comparison` operator. In comparison, the `=` is an `assignment` operator. We use it, for example, when we create or assign variables or objects. In the example below we see both at work.
+
+Likely the most common way in python to select rows is to use a condition inside selection brackets `[]`
+
+```{python}
+filtered_df = df[df['year'] == 1972]
+filtered_df.head(2)
+```
+
+Let's look at what is happening under the hood.  We see that pandas checks each row to see which rows fulfill the condition specified.  
+
+```{python}
+df['year'] == 1972
+```
+
+We can also expand this concept by adding additional conditions that narrow or expand the number of rows filtered. Below we use the `&` condition to indicate that both parts of the equation need to be fulfilled. 
+
+When combining multiple conditional statements as in the example below, each condition must be surrounded by parentheses `()` within the square brackets `[]`. You must use the "or" operator `|` and/or the "and" operator `&` for the code to work. 
+
+```{python}
+filtered_df = df[(df['year'] == 1972) & (df['country'] == 'Albania')]
+filtered_df.head()
+```
+
+In the example above, there is only one row of data that matches the condition.
+
+Or we can use the `|` which indicates that if *either* of the statements are true, then all the rows satisfying either condition are returned.
+
+```{python}
+filtered_df = df[(df['year'] == 1972) | (df['country'] == 'Albania')]
+filtered_df.head()
+```
+
+In the example above, there are many rows that satisfy the condition.
+
+:::{.callout-warning icon="false"}
+
+## Challenge 2
+Your director has come to you and asked if you know what the life expectancy has been in Canada since 1992.  How would you use pandas code to get the data you need? And after having run the code, what's the answer? 
+:::
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 2
+
+First you would like to create a new object to hold this narrowed down version of the dataset, then use some method like `.print()` or `.head()` to see some of the resulting data. 
+
+
+```{python}
+challenge2_df = df[(df['year'] >= 1992) & (df['country'] == 'Canada')]
+print(challenge2_df)
+```
+:::
+
+
+You will find that it is very common within python/pandas to see the `.iloc()` method.
+
+The `.iloc()` method identifies rows based on their index value (that column to the left of the dataframe when you call the <code>.head()</code> function, for example). It can be used to look at individual rows of data by their original index value.
```

Comment:
This explanation of `iloc` is still not quite right. `iloc` doesn't actually refer to the index column at all. I think it makes sense to think of it as there being 2 different ways to index a dataframe:

|  | col 1 | col 2 | col 3 |
| -----  | ----- | ------ | ----- |
| row 1 | | | |
| row 2 | | | |
| row 3 | | | |

VS

|  | 0 | 1 | 2 |
| -----  | ----- | ------ | ----- |
| 0 | | | |
| 1 | | | |
| 2 | | | |

the order of, and names of, rows1-3 and cols1-3 can change, but the second view will always just be 0-2/0-2 no matter how you rearrange and rename the data.

To slice the data using the first view we use `loc`:
df.loc[ row1:row3, col1: col2]

Whereas to slice the data using the second view we use `iloc`:
df.iloc[ 0:-1, 1:2]

Reply:
Arrrgh! Thanks

Reply:
I use both loc and iloc depending on the situation - but off the top of my head I can't think of a specific use case of course. I think if I'm doing a random sampling it's often easier to pick out a random subset of integers than it is row names?

Reply:
Yup, that makes sense for sure - maybe an example of that then would be good?